File: python_team2/ch03/train.py
import sys
sys.path.append('../book')
from common.trainer import Trainer
from common.optimizer import Adam
from ch03.simple_skip_gram import SimpleSkipGram
from simple_cbow import SimpleCBOW
from common.util import preprocess, create_contexts_target, convert_one_hot, most_similar


window_size=1
hidden_size=5
batch_size=3
max_epoch=1000

text  = 'You say goodbye and I say hello.'
#text='Deep learning (also known as deep structured learning or hierarchical learning) is part of a broader family of machine learning methods based on artificial neural networks. Learning can be supervised, semi-supervised or unsupervised. Deep learning architectures such as deep neural networks, deep belief networks, recurrent neural networks and convolutional neural networks have been applied to fields including computer vision, speech recognition, natural language processing, audio recognition, social network filtering, machine translation, bioinformatics, drug design, medical image analysis, material inspection and board game programs, where they have produced results comparable to and in some cases superior to human experts. Artificial Neural Networks (ANNs) were inspired by information processing and distributed communication nodes in biological systems. ANNs have various differences from biological brains. Specifically, neural networks tend to be static and symbolic, while the biological brain of most living organisms is dynamic (plastic) and analog.'

corpus, word_to_id, id_to_word = preprocess(text)

vocab_size = len(word_to_id)
contexts, target = create_contexts_target(corpus, window_size)
target = convert_one_hot(target, vocab_size)
contexts = convert_one_hot(contexts, vocab_size)

model = SimpleCBOW(vocab_size, hidden_size)
#model = SimpleSkipGram(vocab_size, hidden_size)
optimizer = Adam()
trainer = Trainer(model, optimizer)

trainer.fit(contexts, target, max_epoch, batch_size)
trainer.plot()


word_vecs = model.word_vecs
for word_id, word in id_to_word.items():
    print(word, word_vecs[word_id])


most_similar('you', word_to_id, id_to_word, word_vecs)
#most_similar('learning', word_to_id, id_to_word, word_vecs)

File: ruby/ch03/train.rb
# frozen_string_literal: true

require 'adam'
require 'simple_cbow'
require 'trainer'
require 'util'

# window_size = 1
hidden_size = 5
batch_size = 3
max_epoch = 1_000

text = 'You say goodbye and I say hello.'
corpus, word_to_id, id_to_word = preprocess(text)

vocab_size = word_to_id.length
contexts, target = create_contexts_target(corpus, window_size: 1)
contexts = convert_one_hot(contexts, vocab_size)
target = convert_one_hot(target, vocab_size)

model = SimpleCBow.new(vocab_size, hidden_size)
optimizer = Adam.new
trainer = Trainer.new(model, optimizer)

trainer.fit(contexts, target, max_epoch: max_epoch, batch_size: batch_size)
trainer.plot

word_vecs = model.word_vecs

id_to_word.each do |word_id, word|
  printf("%s: %s\n", word, word_vecs[word_id, true].to_a)
end
